Content.Server/Silicons/RobotFenceSystem.cs

```diff
@@ -0,0 +1,183 @@
+﻿using System.Numerics;
+using Content.Server.Power.EntitySystems;
+using Content.Shared.Maps;
+using Content.Shared.Physics;
+using Content.Shared.Power;
+using Content.Shared.Silicons;
+using Content.Shared.Silicons.Bots;
+using Robust.Server.GameObjects;
+using Robust.Shared.Map;
+using Robust.Shared.Map.Components;
+using Robust.Shared.Physics;
+using Robust.Shared.Physics.Systems;
+using Robust.Shared.Timing;
+using Robust.Shared.Utility;
+
+namespace Content.Server.Silicons;
+
+public sealed class RobotFenceSystem : EntitySystem
+{
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly PowerReceiverSystem _powerReceiver = default!;
+    [Dependency] private readonly AppearanceSystem _appearance = default!;
+
+    private const float XOff = -.125f;
+    private const float YOff = -.3f;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<RobotFenceComponent, ComponentInit>(OnInit);
+        SubscribeLocalEvent<RobotFenceComponent, PowerChangedEvent>(OnPowerChanged);
+        SubscribeLocalEvent<RobotFenceComponent, ComponentRemove>(OnRemove);
+    }
+
+    private void OnPowerChanged(Entity<RobotFenceComponent> ent, ref PowerChangedEvent evt)
+    {
+        _appearance.SetData(ent.Owner, RobotFenceVisuals.IsOn, evt.Powered);
+        if (!evt.Powered)
+        {
+            DestroyBeam(ent);
+        }
+        else
+        {
+            UpdateFixture(ent);
+        }
+
+    }
+
+    private void DestroyBeam(Entity<RobotFenceComponent> ent)
+    {
+        while (ent.Comp.BeamEntities.Count > 0)
+        {
+            QueueDel(ent.Comp.BeamEntities.Pop());
+        }
+
+        ent.Comp.BeamSubSteps = 0;
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var curTime = _timing.CurTime;
+
+        var query = EntityQueryEnumerator<RobotFenceComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if(comp.NextUpdate > curTime)
+                continue;
+
+            UpdateFixture((uid, comp));
+
+            comp.NextUpdate += comp.UpdatePeriod;
+        }
+    }
+
```

Comment:
You've doing expensive updates on a fixed timer which is really bad.

Reply:
I can probably update this using collision systems or something...?